adhocteam/Head-Start-TTADP
# Download all activity reports without erroring

## Description of change
This would be part of a larger solution. 

On [the JIRA issue](https://ocio-jira.acf.hhs.gov/browse/TTAHUB-160), SJ commented: 
> - Decide on a safe "limit" on the number of reports to download, and then limit requests to this number from the front end (this could be an error or warning message to the user). Requests could be not served, or broken up into several requests.
> - Investigate if there are performance issues that can be resolved that will make the download feature performant for any request. Resolve those issues.
> - Offload "large download requests" (again we need a definition here) to a worker job. Add a feature (page? inbox?) on the ui where requested downloads are "stored" and their status is shown

This PR only attempts to solve the issue where large downloads crash the site. 

Testing locally, I added a large dataset (10,000+ reports) and verified that downloading all table data crashed my local backend.

I made two changes. 

Firstly, I included the separate:true parameter where appropriate on the activity reports query. This is referenced a few places, including [this aged GH issue](https://github.com/sequelize/sequelize/issues/4868), and we've used it elsewhere in the codebase. It subdivides one-to-many includes into their own queries and joins them in memory.

Secondly, the front end now splits the download in sections of 5000 reports, and the backend has been made to handle it.

## How to test

Use the download links and notice how 5000 reports does not crash your instance.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-160


## Checklists

### Every PR



### Production Deploy


### After merge/deploy
